wallet: batch and simplify addressbook migration process

furszy opened the pull request:
Commits decoupled from #28574, focused on the address book cloning process

Includes:

1) DB batch operations and flow simplification for the address book migration process.
2) Code improvements to `CWallet::DelAddressBook` and `Wallet::SetAddrBookWithDB` methods.

These changes will let us consolidate all individual write operations that take place during the wallet migration process into a single db txn in the future.

furszy committed:
wallet: clean redundancies in DelAddressBook

1) Encode destination only once (instead of three).
2) Fail if the entry's linked data cannot be removed.
3) Don't remove entry from memory if db write fail.
4) Notify GUI only if removal succeeded

furszy committed:
refactor: SetAddrBookWithDB, signal only if write succeeded

w0xlt reviewed (commented):
Approach ACK.
I will review soon.

rserranon commented on src/wallet/wallet.cpp:
```diff
@@ -2394,31 +2394,42 @@ bool CWallet::SetAddressBook(const CTxDestination& address, const std::string& s
 bool CWallet::DelAddressBook(const CTxDestination& address)
 {
-    bool is_mine;
+    isminetype is_mine;
+    const std::string& dest = EncodeDestination(address);
```
Nice optimization!

rserranon reviewed (commented):
Concept ACK, 

I understand this may become a separate commit, but why not encapsulate the other calls to `wallet->m_address_book[addr_pair.first].purpose = purpose`  ?  

```c++
src/wallet/wallet.cpp:3980:                            data.watchonly_wallet->m_address_book[addr_pair.first].purpose = purpose;
src/wallet/wallet.cpp:3996:                            data.solvable_wallet->m_address_book[addr_pair.first].purpose = purpose;
src/wallet/wallet.cpp:4017:                    data.watchonly_wallet->m_address_book[addr_pair.first].purpose = purpose;
src/wallet/wallet.cpp:4030:                    data.solvable_wallet->m_address_book[addr_pair.first].purpose = purpose;
```
If `m_address_book` has been encapsulated, why not make it private or protected?

furszy committed:
refactor: SetAddressBookWithDB, minimize number of map lookups

furszy committed:
refactor: wallet, simplify addressbook migration

Same process written in a cleaner manner.
Removing code duplication.

furszy commented:
Thanks @rserranon.

> why not encapsulate the other calls to wallet->m_address_book[addr_pair.first].purpose = purpose ?

Good catch. I did this changes prior the wallet migration PR merge. Will add those too and make the addressbook member private so we don't keep adding more in the future.

furszy commented on src/wallet/wallet.h:
```diff
@@ -480,6 +480,12 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     //! Adds a destination data tuple to the store, without saving it to disk
     void LoadDestData(const CTxDestination& dest, const std::string& key, const std::string& value) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    //! Adds destination label to the store, without saving it to disk
+    void LoadAddrbookEntryLabel(const CTxDestination& dest, const std::string& label) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
done

furszy commented:
Updated per feedback. Now we no longer access the wallet's addressbook member externally.

Plus, improvements to the address book migration process:
1) Fixed a bug where we don't copy the "send" records to all the wallets.
2) Have re-written the process with no code duplication.
3) Batched db writes so the disk dump is done all at once.

rserranon reviewed (approved):
tAck
* unit tests
* functional tests
* qt & bitcoin-cli manual tests

hernanmarino reviewed (approved):
tested ACK, both unit and functional tests.

pablomartin4btc commented on src/wallet/wallet.cpp:
```diff
@@ -3928,95 +3970,63 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
+    for (const auto& [dest, record] : m_address_book) {
+        // Append "receive" data to other wallets only if not mine.
+        // Labels for everything else ("send") should be cloned to all.
+        bool require_is_mine = record.purpose == "receive" && !IsMine(dest);
+        bool copied = false;
```
nit:
```suggestion
        bool require_is_mine{record.purpose == "receive" && !IsMine(dest)};
        bool copied{false};
```

furszy replied:
I prefer the current code, it's more readable.
The curly braces initializer is primarily useful to prevent narrowing (which is the implicit conversion of arithmetic values). Here, there is no loss of accuracy, we are simply initializing a boolean from a logical statement.

pablomartin4btc commented on src/wallet/wallet.cpp:
```diff
@@ -2394,31 +2410,42 @@ bool CWallet::SetAddressBook(const CTxDestination& address, const std::string& s
 {
-    bool is_mine;
+    isminetype is_mine;
+    const std::string& dest = EncodeDestination(address);
```
```suggestion
    const std::string& dest{EncodeDestination(address)};
```

furszy replied:
Thanks but same as above, I prefer current code as it's more readable and there is no possible narrowing issue here.

pablomartin4btc reviewed (commented):
Approach ACK.
I'll perform some testing soon.

pablomartin4btc reviewed (approved):
tested ACK.
On the code review, went thru the different commits of the PR.
Performed manual testing using `bitcoin-qt`.

furszy commented:
Rebased, conflicts solved.
Plus, per https://github.com/bitcoin/bitcoin/pull/26761#discussion_r1134636537 request, added coverage for wallet address book migration process.

pablomartin4btc reviewed (commented):
re-ACK 8fa0db1d487d76e3fa5ace0bd4d9a8b6bf8dbde1

furszy commented:
rebased post #27224 merge, conflicts solved.

furszy commented:
rebased

furszy commented:
rebased, conflicts solved.

achow101 commented:
This PR ends up doing a lot of different things that are only tangentially related. I think it could be split up into a few smaller PRs, e.g. one for the migration fix, one for the migration refactoring, and one for the other refactors.

furszy reviewed (commented):
> This PR ends up doing a lot of different things that are only tangentially related. I think it could be split up into a few smaller PRs, e.g. one for the migration fix, one for the migration refactoring, and one for the other refactors.

sure, will work on it.

furszy committed:
wallet: migration, batch addressbook records removal

Instead of doing one db transaction per removed record,
we now batch all removals in a single db transaction.

Speeding up the process and preventing the wallet from entering
an inconsistent state when any of the intermediate writes fail.

furszy committed:
wallet: addressbook migration, batch db writes

Optimizing the process performance and consistency.

furszy reviewed (commented):
In response to feedback, the PR scope has been shortened, omitting the address book encapsulation commits. The current version includes only the subset of commits related to the address book from #28574. Decoupling this PR into two is also possible upon request.

josibake commented on src/wallet/wallet.cpp:
```diff
@@ -3999,52 +3999,41 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
-                // Add to the solvable. Copy the entire address book entry
-                data.solvable_wallet->m_address_book[addr_pair.first] = addr_pair.second;
+            // Skip invalid/non-watched scripts that will not be migrated
+            if (not_migrated_dests.count(dest) > 0) {
```
in "refactor: wallet, simplify addressbook migration" (https://github.com/bitcoin/bitcoin/pull/26836/commits/8b331613bf79ab614850e71a603dd444d6bb3a48):

nit: C++20 has `.contains(arg)` for sets (yay!). No performance difference, but semantically I think it makes a lot more sense to use when testing for existence.

josibake reviewed (approved):
crACK https://github.com/bitcoin/bitcoin/pull/26836/commits/3aff6d5b4f3fd32ad1285b259db0ef1580518e39

Looks good! Code reviewed, didn't verify the bench numbers. Left some nits, feel free to ignore

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3999,52 +3999,41 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
+            // Only delete 'receive' records that are no longer part of the original wallet
+            if (require_transfer) {
+                dests_to_delete.push_back(dest);
+                break;
```
> If the destination requires_transfer but isn't ours, go to the next wallet. If this happens for both wallets, copy will never be set and we will fail

If the destination requires transfer (a "receive" addr that is no longer part of the main wallet), the process checks to which wallet the entry belongs to. And, when the entry does not belong to any of the created wallets, it means that the key/script-to-descriptor migration process failed. This is because "receive" entries must always be tracked by the wallet.

> If it requires transfer and does belong to the current wallet in the loop, continue and set copy = true, and add the destination to dests_to_delete. Do not check the next wallet. I'm assuming the assumption here is it can only belong to one of the wallets, and if it does end up happening to belong to both wallets this is not an issue?

Yes, thats the existing assumption. If you check the previous code, will see that I did not change this part of the flow on purpose. This is what we had before:

```c++
if (purpose == "RECEIVE" && not_mine_anymore) {

    // check if it is part of the watch-only wallet
    if (watch_only_wallet.IsMine(dest)) {
        store_address(watch_only_wallet, dest);
        add_to_pending_to_delete(dest);
        continue; // --> this skips the next wallet.
    } 

    // check if it is part of the solvables wallet
    if (solvables_wallet.IsMine(dest)) {
        store_address(solvables_wallet, dest);
        add_to_pending_to_delete(dest);
        continue; // --> skips any following-up check
    } 

    // We only reach here if the entry wasn't added to any wallet
    error out...
}

```

The only diff between the new code and the previous one is that we had a `continue` instead of a `break` statement.


 > Might be worth adding a comment explaining why it's okay to break early? That's the part that kinda threw me.

We should double-check if we could ever have the same no-longer ours "receive" entry in both wallets (the watch-only and the solvables ones). If we find a way to achieve this, then the bug predates this PR.

furszy reviewed (commented):
Updated per feedback, thanks for the review josie!

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -2394,8 +2394,13 @@ bool CWallet::SetAddressBook(const CTxDestination& address, const std::string& s
+    return DelAddressBookWithDb(batch, address);
+}
+
+bool CWallet::DelAddressBookWithDb(WalletBatch& batch, const CTxDestination& address)
```
sure

furszy reviewed (commented):
Updated per @maflcko review in https://github.com/bitcoin/bitcoin/pull/29403#discussion_r1481506563. [Small diff](https://github.com/bitcoin/bitcoin/compare/238993ea40c1c384e9fbcf83bfcf45efc73ad1bb..4c4c62b459e5b92392a67cb369101b07fd4bad28).

furszy commented:
Green CI. Ready to go.

achow101 reviewed (commented):
ACK 1f177ff9a6ab229bd6486941e46daa92ab22b622

josibake commented:
reACK https://github.com/bitcoin/bitcoin/pull/26836/commits/1f177ff9a6ab229bd6486941e46daa92ab22b622

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -2368,13 +2368,22 @@ bool CWallet::SetAddressBookWithDB(WalletBatch& batch, const CTxDestination& add
+        return false;
+    }
+    if (!batch.WriteName(encoded_dest, strName)) {
+        WalletLogPrintf("%s error writing entry name\n", __func__);
```
Sure. `__func__ ` removed.

furszy reviewed (commented):
Updated per feedback. Thanks achow!. [Small diff](https://github.com/bitcoin/bitcoin/compare/1f177ff9a6ab229bd6486941e46daa92ab22b622..86960cdb7f75eaa2ae150914c54240d1d5ef96d1).

Two changes:
1) Removed `__func__ ` usages in the logging messages ([comment_1](https://github.com/bitcoin/bitcoin/pull/26836#discussion_r1482012171) and [comment_2](https://github.com/bitcoin/bitcoin/pull/26836#discussion_r1482012269)).
2) Have divided the purpose and name erasing lines per [request](https://github.com/bitcoin/bitcoin/pull/29403#discussion_r1481997602).

ziljah commented:
👎

achow101 commented:
ACK 86960cdb7f75eaa2ae150914c54240d1d5ef96d1

josibake commented:
reACK https://github.com/bitcoin/bitcoin/commit/86960cdb7f75eaa2ae150914c54240d1d5ef96d1

Verified the diff, looks good.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4009,52 +4009,41 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
+    for (const auto& [dest, record] : m_address_book) {
+        // Ensure "receive" entries that are no longer part of the original wallet are transferred to another wallet
+        // Entries for everything else ("send") will be cloned to all wallets.
+        bool require_transfer = record.purpose == AddressPurpose::RECEIVE && !IsMine(dest);
```
In commit "refactor: wallet, simplify addressbook migration" (595bbe6e81885d35179aba6137dc63d0e652cc1f)

IMO, this would be less confusing if `require_transfer` bool was replaced by an `copy_to_all` bool with the opposite meaning, because the thing `require_transfer` is used for in the loop below is to determining whether the address book entry is copied to all wallets or is moved from the `*this` wallet to one of the other two.

The name `require_transfer` is also misleading because the transfer is not required if `dest` is in `not_migrated_dests`.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4007,21 +4007,44 @@ bool CWallet::ApplyMigrationData(MigrationData& data, bilingual_str& error)
+    }
+
+    // Write address book entry to disk
+    auto func_store_addr = [](WalletBatch& batch, const CTxDestination& dest, const CAddressBookData& entry) {
```
In commit "wallet: addressbook migration, batch db writes" (342c45f80e32b0320829ce380b5854844cd74bc8)

`write_address_book` would be a more descriptive name than `func_store_addr` and `watchonly_wallets` would be more descriptive than `wallet_vec`. The `_vec` suffix and and `func_` prefix do not do anything to help a reader of the code, IMO.

ryanofsky reviewed (commented):
Post-merge code review ACK 86960cdb7f75eaa2ae150914c54240d1d5ef96d1. Nice code cleanups, and use of batches to speed things up
